sycl/doc/extensions/experimental/sycl_ext_oneapi_group_occupancy_queries.asciidoc

```diff
@@ -0,0 +1,154 @@
+specification.*
+
+
+== Overview
+
+This extension is based on the kernel-queue-specific specific querying mechanism
+introduced by the sycl_ext_oneapi_launch_queries extension.
+
+The purpose of queries the to be added is to aid occupancy based calculations
+for kernel launches based on hardware occupancy per compute unit granularity.
+The queries take in account the kernel resources and user-specified constraints,
+such as, but not limited to, local (work-group) size and dynamic work-group
+local memory (in bytes). The motivation behind is to aid the tuning of kernels,
+by being able to design the algorithm's implementation to maintain the highest
+possible occupancy in a portable way.
+
+List of currently planned queries.
+* max_num_work_group_occupancy_per_cu
+
+[source,c++]
+----
+sycl::queue q{};
+auto bundle = sycl::get_kernel_bundle(q.get_context());
+auto kernel = bundle.get_kernel<class KernelName>();
+
+auto wgSizeRange = sycl::range{32, 1, 1};
+size_t localMemorySize = 32;
+
+namespace syclex = sycl::ext::oneapi::experimental;
+uint32_t maxWGsPerCU = kernel.ext_oneapi_get_info<
+      syclex::info::kernel_queue_specific::max_num_work_group_occupancy_per_cu>(
+      q, wgSizeRange, localMemorySize);
+----
+
+NOTE: SYCL 2020 requires lambdas to be named in order to locate the associated
+`sycl::kernel` object used to query information descriptors. Reducing the
+verbosity of the queries shown above is left to a future extension.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_GROUP_OCCUPANCY_QUERIES` to one of the values defined in
+the table below. Applications can test for the existence of this macro to
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's features the
+implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== Occupancy queries
+
+[source, c++]
+----
+namespace ext::oneapi::experimental::info::kernel {
+
+struct max_num_work_group_occupancy_per_cu;
+
+}
+----
+
+[%header,cols="1,5,5,5"]
+|===
+|Kernel Descriptor
+|Argument Types
+|Return Type
+|Description
+
+|`max_num_work_group_occupancy_per_cu`
+|`sycl::queue`, `sycl::range`, `size_t`
+|`uint32_t`
+|Returns the maximum number of actively executing work-groups per compute unit
+granularity, when the kernel is submitted to the specified queue with the
+specified work-group size and the specified amount of dynamic work-group local
+memory (in bytes). The actively executing work-groups are those that occupy
+the fundamental hardware unit responsible for the execution of work-groups in
+parallel.
```

Comment:
Is the idea that `max_num_work_group_occupancy_per_cu` returns a recommended work-group size?  If that is the case, can we rename the query to something like `recommended_num_work_groups`?

Reply:
It is a recommendation for the maximum number of work-groups (or Cuda blocks) of specified block size etc. that will theoretically execute concurrently on the compute unit (Cuda SM) to achieve maximum occupancy.

I think I like the naming you suggest and it does sound to me like a recommendation. What do you think would be a good name based on my description? (I am sold on the fact the original I came up with sounds a little weird.)

Thank you, @gmlueck !

Reply:
I renamed it to `recommended_num_work_groups`. Initially, I wanted to indicate that this is not per-device semantics but per compute unit (or whatever this maps to in the HW, i.e. SM for Cuda, EU for Intel Level-Zero or CU for AMD HIP), hence why I had the `_per_cu` in the name. However, the extension docs describe the semantics, so I think that's okay now.